torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -177,6 +177,49 @@ struct DeviceTypePropagationPass {
     }
   }
 
+  bool mergeAndApplyTensorProps(
```

Comment:
can we just use `mergeTypes` here from `shape_analysis.h`?

Reply:
We can, but I thought we wanted to kill off `shape_analysis.h / shape_analysis.cpp`

Reply:
`default to union` doesn't affect what we are doing here. 
```
// If no supertype is found, then nullopt will be returned if
// `default_to_union` is false, and `Union[t1, t2]` will be returned
// if it is true. If `t1 == t2`, or `t1` is a type refinement of `t2`,
// then `t2` will be returned (and vice versa).
```

That api (true or false) doesn't affect what we're doing here, which is just adding type refinements to TensorTypes